### PR TITLE
Migrate Refactoring for Generating Equal and Hash Methods

### PR DESCRIPTION
**Pull request by hernanmd**
# Description

This PR migrates the refactoring browser for generating #= and #hash methods to the new driver interface. The new interface provides a more streamlined and extensible approach for generating these methods, making it easier to maintain and enhance the functionality in the future.

# Motivation

The existing implementation of the refactoring browser for generating #= and #hash methods was tightly coupled with the user interface (UI) code and did not prevent overwriting the existing implementation of such methods (#16361). By separating the core logic from the UI, we can improve the code's maintainability and testability and enable future enhancements or alternative user interfaces.

# Implementation.

The implementation of this PR involves the following steps:

- Create a new driver interface: A new driver interface, `ReGenerateEqualAndHashDriver,` has been introduced. This interface defines the contract for generating #= and #hash methods, encapsulating the core logic for this functionality.
- Implement the driver: The driver uses `RBGenerateEqualHashTransformation` for the underlying logic.
- UI interaction: The driver uses `SpSelectMultipleDialog` for selecting the variables used in the generated #= and #hash methods.
- Created tests: Tests related to the generation of #= and #hash methods have been provided in `ReGenerateEqualAndHashDriverTest`

### CODE REVIEW

**Comment by Ducasse** [comment]
@balsa-sarenac can you have a look? Tx

File: src/Refactoring-Core/RBGenerateEqualHashTransformation.class.st [hunk 1]
```diff
@@ -88,8 +88,7 @@ RBGenerateEqualHashTransformation >> accessorForVariable: aString [
 RBGenerateEqualHashTransformation >> applicabilityPreconditions [
```
Comment (Ducasse) [inline src/Refactoring-Core/RBGenerateEqualHashTransformation.class.st]: Can you rename the class Re?

File: src/Refactoring-DataForTesting/RBClassForGeneratingEqualAndHash.class.st [hunk 2]
```diff
@@ -0,0 +1,11 @@
+Class {
+	#name : 'RBClassForGeneratingEqualAndHash',
```
Comment (Ducasse) [inline src/Refactoring-DataForTesting/RBClassForGeneratingEqualAndHash.class.st]: All new classes or migrated classes should start with Re

File: src/Refactoring-DataForTesting/RBClassForGeneratingEqualAndHashExistingImplementors.class.st [hunk 3]
```diff
@@ -0,0 +1,27 @@
+Class {
+	#name : 'RBClassForGeneratingEqualAndHashExistingImplementors',
```
Comment (Ducasse) [inline src/Refactoring-DataForTesting/RBClassForGeneratingEqualAndHashExistingImplementors.class.st]: Re

File: src/Refactoring-UI-Tests/ReGenerateEqualAndHashDriverTest.class.st [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	#category : 'Refactoring-UI-Tests-Driver',
+	#package : 'Refactoring-UI-Tests',
+	#tag : 'Driver'
+}
+
+{ #category : 'tests' }
+ReGenerateEqualAndHashDriverTest >> classForEqualAndHash [
+	"Answer the <Class> which will be used to generate equal and hash methods"
+	
+	^ RBClassForGeneratingEqualAndHash
```
Comment (Ducasse) [inline src/Refactoring-UI-Tests/ReGenerateEqualAndHashDriverTest.class.st]: Re

File: src/SystemCommands-ClassCommands/SycCmGenerateEqualAndHashCommand.class.st [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+{ #category : 'testing' }
+SycCmGenerateEqualAndHashCommand >> isApplicable [
+	"Answer <true> if the receiver has any variables"
+
+	^ context lastSelectedClass realClass slots notEmpty
+]
+
+{ #category : 'accessing' }
+SycCmGenerateEqualAndHashCommand >> name [
```
Comment (Ducasse) [inline src/SystemCommands-ClassCommands/SycCmGenerateEqualAndHashCommand.class.st:45]: The name of migrated refactoring should have (R) in front 
the name of migrated transformation should have (T) in front.

File: src/SystemCommands-ClassCommands/SycCmGenerateEqualAndHashCommand.class.st [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+{ #category : 'accessing' }
+SycCmGenerateEqualAndHashCommand >> order [
+	^ 41
+]
+
+{ #category : 'accessing' }
+SycCmGenerateEqualAndHashCommand >> refactoringClass [
+
+	^ ReGenerateEqualHashTransformation
+]
```
Comment (balsa-sarenac) [inline src/SystemCommands-ClassCommands/SycCmGenerateEqualAndHashCommand.class.st]: I think we don't need this anymore since we're using the driver now

**Review by balsa-sarenac — APPROVED** [review]

File: src/SystemCommands-ClassCommands/SycCmGenerateEqualAndHashCommand.class.st (hunk 6 above)
Comment (hernanmd) [reply src/SystemCommands-ClassCommands/SycCmGenerateEqualAndHashCommand.class.st]: You mean to implement `ReGenerateEqualAndHashDriver>>refactoringClass` to answer ReGenerateEqualHashTransformation?

**Comment by Ducasse** [comment]
Yes

**Comment by Ducasse** [comment]
Sounds good to me. 
Potentially the other menuString should be revisited. 
But I will integrate the PR right now.

File: src/SystemCommands-ClassCommands/SycCmGenerateEqualAndHashCommand.class.st (hunk 6 above)
Comment (balsa-sarenac) [reply src/SystemCommands-ClassCommands/SycCmGenerateEqualAndHashCommand.class.st]: I mean remove it completely. I didn't saw any usages?

Comment (hernanmd) [reply src/SystemCommands-ClassCommands/SycCmGenerateEqualAndHashCommand.class.st]: There is usage in `SycCMakeAbstractCommand,` `SycRealizeClassCommand`, and `SycGeneratePrintOnCommand.` However, there should be a way to get the `menuItemString`

**Comment by hernanmd** [comment]
> Sounds good to me. Potentially the other menuString should be revisited. But I will integrate the PR right now.

Thanks

**Comment by demarey** [comment]
@hernanmd The renaming is not good: `Re`prefix si used by Renraku classes (critics).
Also, we now have some release tests failing :
``` 
the following classes have unused instance variables and should be cleaned:ReClassForGeneratingEqualAndHashExistingImplementors ReClassForGeneratingEqualAndHash

ReleaseTest.testThatThereAreNoSelectorsRemainingThatAreSentButNotImplemented
 'ReClassForGeneratingEqualAndHashExistingImplementors>>#=
```

**Comment by hernanmd** [comment]
> ReClassForGeneratingEqualAndHashExistingImplementors

Thanks for reporting, Christophe.
Would you like me to fix it in another PR?

We should probably discuss the Re prefix (as taken by Renraku) with @Ducasse.

**Comment by hernanmd** [comment]
> ReClassForGeneratingEqualAndHashExistingImplementors

Ok, I checked. Those instance variables are on purpose, so I added a Manifest.
(I cannot re-open the PR)

**Comment by demarey** [comment]
> > ReClassForGeneratingEqualAndHashExistingImplementors
> 
> Ok, I checked. Those instance variables are on purpose, so I added a Manifest. (I cannot re-open the PR)

Thank you for checking Hernan!

**Comment by demarey** [comment]
> > ReClassForGeneratingEqualAndHashExistingImplementors
> 
> Thanks for reporting, Christophe. Would you like me to fix it in another PR?
> 
> We should probably discuss the Re prefix (as taken by Renraku) with @Ducasse.

yes, I think it would be better to use another prefix. When you agree on a prefix, you can issue a new PR for it.

**Comment by Ducasse** [comment]
Why this is a problem to have two projects with the same prefix. 
We could have ren for renraku and re for refactoring engine

**Comment by demarey** [comment]
Having the same prefix for different projects just add confusion even if technically, it does not change anything.
People are used to identify the project of a class by checking the prefix

**Comment by MarcusDenker** [comment]
This PR was merged but it had failing tests, among them:

- [CI] failing test: testNoUnusedInstanceVariablesLeft #16824
- [CI] testThatThereAreNoSelectorsRemainingThatAreSentButNotImplemented #16825

**Comment by MarcusDenker** [comment]
I sadly can not even revert the PR as this is now so long ago that it was merged

**Pull request merged**